scene/gui/layout_container.h

```diff
@@ -0,0 +1,122 @@
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef LAYOUT_CONTAINER_H
+#define LAYOUT_CONTAINER_H
+
+#include "core/math/math_defs.h"
+#include "scene/gui/container.h"
+
+class LayoutContainer : public Container {
+	GDCLASS(LayoutContainer, Container);
+
+public:
+	enum LayoutContainerPreset {
+		PRESET_TOP_LEFT = 0,
+		PRESET_TOP_RIGHT,
+		PRESET_BOTTOM_LEFT,
+		PRESET_BOTTOM_RIGHT,
+		PRESET_CENTER_LEFT,
+		PRESET_CENTER_TOP,
+		PRESET_CENTER_RIGHT,
+		PRESET_CENTER_BOTTOM,
+		PRESET_CENTER,
+		PRESET_LEFT_WIDE,
+		PRESET_TOP_WIDE,
+		PRESET_RIGHT_WIDE,
+		PRESET_BOTTOM_WIDE,
+		PRESET_VCENTER_WIDE,
+		PRESET_HCENTER_WIDE,
+		PRESET_FULL_RECT,
+		PRESET_CUSTOM,
+		PRESET_MAX,
+	};
+
+	const real_t relative_margins_presets[PRESET_CUSTOM][4] = {
+		// LEFT, TOP, RIGHT, BOTTOM.
+		{ 0.0, 0.0, 1.0, 1.0 }, // PRESET_TOP_LEFT
+		{ 1.0, 0.0, 0.0, 1.0 }, // PRESET_TOP_RIGHT
+		{ 0.0, 1.0, 1.0, 0.0 }, // PRESET_BOTTOM_LEFT
+		{ 1.0, 1.0, 0.0, 0.0 }, // PRESET_BOTTOM_RIGHT
+		{ 0.0, 0.5, 1.0, 0.5 }, // PRESET_CENTER_LEFT
+		{ 0.5, 0.0, 0.5, 1.0 }, // PRESET_CENTER_TOP
+		{ 1.0, 0.5, 0.0, 0.5 }, // PRESET_CENTER_RIGHT
+		{ 0.5, 1.0, 0.5, 0.0 }, // PRESET_CENTER_BOTTOM
+		{ 0.5, 0.5, 0.5, 0.5 }, // PRESET_CENTER
+		{ 0.0, 0.0, 1.0, 0.0 }, // PRESET_LEFT_WIDE
+		{ 0.0, 0.0, 0.0, 1.0 }, // PRESET_TOP_WIDE
+		{ 1.0, 0.0, 0.0, 0.0 }, // PRESET_RIGHT_WIDE
+		{ 0.0, 1.0, 0.0, 0.0 }, // PRESET_BOTTOM_WIDE
+		{ 0.5, 0.0, 0.5, 0.0 }, // PRESET_VCENTER_WIDE
+		{ 0.0, 0.5, 0.0, 0.5 }, // PRESET_HCENTER_WIDE
+		{ 0.0, 0.0, 0.0, 0.0 }, // PRESET_FULL_RECT
+	};
+
+private:
+	LayoutContainerPreset preset = PRESET_TOP_LEFT;
+	real_t custom_relative_margins[4] = { ANCHOR_BEGIN, ANCHOR_BEGIN, ANCHOR_BEGIN, ANCHOR_BEGIN };
+	real_t margins[4] = { 0.0, 0.0, 0.0, 0.0 };
+
+	struct ThemeCache {
```

Comment:
I wonder if there is any use in having margins as theme property, and duplicating regular property 🤔
MarginContainer has margins in theme, but I have never seen them used outside overrides.

Reply:
I don't really know. I think it's useful to keep compatibility, but maybe I should only keep the theme ones ? 
I kind of find annoying to have to set them there though, the theme overrides are a bit buried down the tree. :/
So well, I kind of find having both useful, but maybe we should find another solution IDK.

Reply:
But what are theme values useful for? Theme values are for something that you set in Theme. If they are only used via overrides then they should be regular properties.

> I think it's useful to keep compatibility

Compatibility in a new class?

Reply:
> Compatibility in a new class?

Compatibility with MarginContainer. Which this class could replace if we keep those theme properties.

Reply:
Yeah, but as I said above, in MarginContainer they are useless as theme property. I never seen or used them in a Theme, only as override, because you never want **all** your MarginContainers to have the same margins. So having a theme property in addition to a normal property is just redundant.

> I kind of find annoying to have to set them there though, the theme overrides are a bit buried down the tree. :/

https://github.com/godotengine/godot-proposals/issues/4274 would solve that.

Reply:
> Yeah, but as I said above, in MarginContainer they are useless as theme property. I never seen or used them in a Theme, only as override, because you never want all your MarginContainers to have the same margins. So having a theme property in addition to a normal property is just redundant.

Well, my assumption would be that you could use MarginContainers with themes to make everything more breathable (not a single component). At least it would explain why those were theme constants in the first place. So, I'd rather make sure those are never used in themes.

Of course, if almost everyone uses them as overrides, we could remove them. Though I'm not sure we could consider LayoutContainer as a drop-in replacement for MarginContainer anymore.